src/pygeoprocessing/routing/routing.pyx

```diff
@@ -850,34 +860,45 @@ def fill_pits(
                 # if it's a pit or plateau
                 search_queue.push(CoordinateType(xi_root, yi_root))
                 flat_region_mask_managed_raster.set(xi_root, yi_root, 1)
-                downhill_drain = 0
-                nodata_drain = 0
+                natural_drain_exists = 0
 
                 # this loop does a BFS starting at this pixel to all pixels
                 # of the same height. the _drain variables are used to
                 # remember if a drain was encountered. it is preferable to
                 # search the whole region even if a drain is encountered, so
                 # it can be entirely marked as processed and not re-accessed
                 # on later iterations
+                search_steps = 0
                 while not search_queue.empty():
                     xi_q = search_queue.front().xi
                     yi_q = search_queue.front().yi
                     search_queue.pop()
+                    if search_steps > max_pixel_fill_count:
+                        # clear the search queue and quit
+                        while not search_queue.empty():
+                            search_queue.pop()
+                        natural_drain_exists = 1
+                        break
+                    search_steps += 1
 
                     for i_n in range(8):
                         xi_n = xi_q+D8_XOFFSET[i_n]
                         yi_n = yi_q+D8_YOFFSET[i_n]
                         if (xi_n < 0 or xi_n >= raster_x_size or
                                 yi_n < 0 or yi_n >= raster_y_size):
-                            nodata_drain = 1
+                            natural_drain_exists = 1
                             continue
                         n_height = filled_dem_managed_raster.get(
                             xi_n, yi_n)
                         if _is_close(n_height, dem_nodata, 1e-8, 1e-5):
-                            nodata_drain = 1
+                            natural_drain_exists = 1
                             continue
                         if n_height < center_val:
-                            downhill_drain = 1
+                            natural_drain_exists = 1
+                            continue
+                        if flat_region_mask_managed_raster.get(

```

Comment:
Just a clarifying question. Can `flat_region_mask_managed_raster.get(xi_n, yi_n)` be anything but 1 or `nodata`? Here we skip if it's been set and identified before, but on 903,904 it's checked again against `mask_nodata`. Just seeing if that redundancy is necessary?

Reply:
It's not necessary! I missed that I'd checked it below. Removed in 485c3f6f19bb031b60912f741543db62d80c6c0d